### PR TITLE
initial readme

### PR DESCRIPTION
**Pull request by mccanne**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1 +1,262 @@
...
+on localhost.
+
+## Ray
+
+Ray provides a distributed computation model for python where you can create
+stateful actors and run stateless tasks that all coordinate to carry out
+distributed computation.  Actors store state as
+immutable objects that other actors and tasks can access.  Actors and
+tasks can launch other actors and tasks and the Ray scheduler and distributed
+object manager make everything work magically together.
```
Comment (alfred-landrum) [inline README.md]:
```suggestion
immutable objects that other actors and tasks can access.  Actors and
tasks can launch other actors and tasks and the Ray scheduler and distributed
object manager coordinate activity across the cluster.
```

File: README.md [hunk 2]
```diff
@@ -1 +1,262 @@
...
+This is all incredibly inefficient because all
+the data is passed through the plasma object store --- this is just meant as a first cut
+here to start brainstorming.  A better approach would be to have a data reader
+efficiently read the filtered results straight into memory that could then be
+operated upon before being resulted into the plasma object store in the Arrow format.
+
+## incremental deployment model
+
+Supposing you had an existing deployment of stored data that was accessed by Ray,
+you could conceivably index this existing data with zar and add a search capability
```
Comment (alfred-landrum) [inline README.md]:
```suggestion
If you have an existing set of data accessible by Ray,
you could conceivably index this data with zar and add a search capability
```

File: README.md [hunk 3]
```diff
@@ -1 +1,262 @@
...
+to any jobs that traverse the stored data file by file using the zar-find actor
+from this example.  In fact, zar find could be parallelized by providing
+a different zar root to search for the set of files under that root that match.
```
Comment (alfred-landrum) [inline README.md]: `In fact, zar find ...` - I would drop this sentence, as it describes a particular sub-optimization for 'zar find' that distracts from the overall goal of the paragraph to introduce the concept of using indexes. The next sentence (`Then, any existing code ...`) still works without it.

File: README.md [hunk 4]
```diff
@@ -1 +1,262 @@
...
+reduceByKey operator with an aggregation method instead of the standard
+groupBy operator to cause partial aggregations to be computed at the
+output stage of spark's map phase, a so called map-combine-reduce pattern.
+
+So, let's talk about a hypothetical search query over some implied time range:
+```
+blah | sum(foo) by bar
+```
+This searches for all data that matches the pattern `blah` and
+sumsa the values of field `foo` with respect to each value in the field `bar`.
```
Comment (alfred-landrum) [inline README.md]:
```suggestion
sums the values of field `foo` with respect to each value in the field `bar`.
```

File: README.md [hunk 5]
```diff
@@ -1 +1,262 @@
...
+But what if we focused on zql-style queries leveraging the typical case of
+composable aggregations and integrate into the design the use of search indexes
+during query traversals?  And what if this were all exposed as simple
+unix tools that can be experimented with from the command-line without
+setting up complex clusters?
+
+### version 1: a static model
+
+The hadoopy approach for all this is to just spill to disk.  Any aggregation
+(or sort etc)
```
Comment (alfred-landrum) [inline README.md]:
```suggestion
(or sort, or other grouping operation)
```

File: README.md [hunk 6]
```diff
@@ -1 +1,262 @@
...
+traversal is divided amongst the different nodes and the final step merges
+the streamed results from all of the nodes.
+
+That will work fine for many use cases.
+For many aggregations, the group-by table will stay small when the cardinality
+of the key space is low.
+If you have a feel for how much memory you need ahead of time, you could
+launch N jobs in parallel so a high-cardinality group-by table
+gets split out efficiently.  In this case, the key space isn't disjoint
+but it might not matter much if the sub-cardinilaty of each key is small
```
Comment (alfred-landrum) [inline README.md]:
```suggestion
but it might not matter much if the sub-cardinality of each key is small
```

File: README.md [hunk 7]
```diff
@@ -1 +1,262 @@
...
+(meaning there is little advantage to a shuffle)
+so in general this could work well depending on the underlying data
+characteristic.  And if the memory limit is hit here, the table is spilled
+to disk as above.
+
+### version 2: a dynamic model
+
+There are two key problems with the static approach.  First, you don't generally
+know how to choose N so it's unreasonable to ask the user to specify this.
+And second, the key space is replicated over each node making the tables inefficient
```
Comment (alfred-landrum) [inline README.md]:
```suggestion
And second, partitioning the work by time means the key space is replicated over each node making the tables inefficient
```

File: README.md [hunk 8]
```diff
@@ -1 +1,262 @@
...
+and larger than they need to be.
+
+A solution to these problems is, firstly, to fork a new worker only after the
+aggregation table gets too big (thus providing scale-out memory along with
+scale-out throughput) and, secondly, to implement a shuffle to
+route the table rows to the worker that owns the row.  Ownership of rows
+is handled like spark and map-reduce using a hash partition.  The
+number of dynamics nodes would be limited either because the cluster
+manager (i.e., kubernetes) indicates no more resources are available
```
Comment (alfred-landrum) [inline README.md]:
```suggestion
manager indicates no more resources are available
```

Comment (mccanne) [reply README.md]: I'm going to change i.e. to e.g. and leave this to imply that we don't need a custom cluster manager like ray and spark have.

File: README.md [hunk 9]
```diff
@@ -1 +1,279 @@
...
+dynamically add workers (increasing N) to provide additional memory scale-out
+and throughput parallelism.  The traversal range is divided among the M workers
+in some way (i.e., evenly divided, partitioned with a block/M stride, etc)
+
+When a worker hits a memory limit because its aggregation table gets to big,
+it splits its job, say in 2 (though this fanout can be configured), or it resorts
+to spilling if no resource is available for the split.  When it splits,
+half of the key space and half of the input partition goes to the new child.
+
+Once split, the worker initiate a _continuous shuffle_, where each worker pushes
```
Comment (alfred-landrum) [inline README.md]:
```suggestion
Once split, the worker initiates a _continuous shuffle_, where it pushes
```

File: README.md [hunk 10]
```diff
@@ -1 +1,279 @@
...
+
+In this approach, we begin with N=M workers for throughput parallelism then
+dynamically add workers (increasing N) to provide additional memory scale-out
+and throughput parallelism.  The traversal range is divided among the M workers
+in some way (i.e., evenly divided, partitioned with a block/M stride, etc)
+
+When a worker hits a memory limit because its aggregation table gets to big,
+it splits its job, say in 2 (though this fanout can be configured), or it resorts
+to spilling if no resource is available for the split.  When it splits,
+half of the key space and half of the input partition goes to the new child.
```
Comment (alfred-landrum) [inline README.md:246]: It would be good to clarify if/how the keyspace is partitioned across the initial M workers.

File: README.md [hunk 11]
```diff
@@ -1 +1,279 @@
...
+
+To optimize the continuous shuffle, each worker keeps N-1 _peer tables_ to batch
+the streaming updates.  As records arrive in this table, they can be aggregated
+before transmission potentially reducing the overhead.  In general,
+larger peer tables results in lower overhead.  A configuration
+parameter controls the ratio of the size of the peer table storage to the
+primary table.  Note that the peer tables do not need to all be the same size and
+can shared a dynamic size to better optimize non-uniform data patterns.
+
+At end-of-stream, a worker flushes its caches and streams back its
```
Comment (alfred-landrum) [inline README.md]:
```suggestion
At end-of-stream, a worker flushes its peer table contents and streams back its
```

File: README.md [hunk 12]
```diff
@@ -1 +1,279 @@
...
+table (perhaps mixing in any spills) to its parent in response to the
+REST call that invoked that worker.  This process of
+writing results may block if the parent is not yet done.  If so, when the
+parent is done and ready, it will read the results from all of its children,
+and propagate the merged results in sorted order to its parent and so forth.
+
+The state required for reach worker to carry all this out is simply the list
+of endpoints that exist and any point in time along with the key space
+partitioning strategy.  Because of the properties of decomposable aggregations,
```
Comment (alfred-landrum) [inline README.md:271]:
```suggestion
partitioning strategy.  Because of the properties of composable aggregations,
```

Comment (mccanne) [reply README.md:271]: We can change the name but I got this term here: 
https://arxiv.org/pdf/1110.0725.pdf

**Review by alfred-landrum — APPROVED** [review]

**Pull request merged**